### PR TITLE
Add python benchmark

### PR DESCRIPTION
**Pull request by BSBandme**
Add python benchmark.

Testing with pure python, cpp reflection, cpp generated code

### CODE REVIEW

**Comment by BSBandme** [comment]
Will update readme after #4014 added

File: benchmarks/py_benchmark.py [hunk 1]
```diff
@@ -0,0 +1,123 @@
...
+def parse_from_benchmark():
+  global counter, message_class, benchmark_dataset
+  message_class().ParseFromString(benchmark_dataset.payload[counter % len(benchmark_dataset.payload)])
+  counter = counter + 1
+  
+def serialize_to_benchmark():
+  global counter, message_list, message_class
+  message_list[counter % len(benchmark_dataset.payload)].SerializeToString()
+  counter = counter + 1
+
```
Comment (BSBandme) [inline benchmarks/py_benchmark.py]: Any other benchmark method need to test?

**Comment by xfxyjwf** [comment]
@anandolee Jie, can you take a look at this change?

**Comment by xfxyjwf** [comment]
@BSBandme Can you post the benchmark result for python? (including C++/Java results as well for comparison).

**Comment by BSBandme** [comment]
Benchmark result:
https://docs.google.com/document/d/1KK_bKVB-34KmVJNTpuncXHXtQq0rYM0tkJ1taAnkLtY/edit?usp=sharing

**Comment by BSBandme** [comment]
benchmark result updated:

## <b>Parsing</b>
<table>
<tr>
<th rowspan="2"> </th>
<th colspan="6" rowspan="1">cpp</th>
<th colspan="6" rowspan="1">java</th>
<th colspan="6" rowspan="1">python</th>
</tr>
<tr>
<th colspan="2">new</th>
<th colspan="2">newarena</th>
<th colspan="2">reuse</th>
<th colspan="2">deserializeFromByteArray</th>
<th colspan="2">deserializeFromByteString</th>
<th colspan="2">deserializeFromMemoryStream</th>
<th colspan="2">cpp-generated-code</th>
<th colspan="2">cpp-reflection</th>
<th colspan="2">pure-python</th>
</tr>
<tr>
<td>google_message1_proto2</td>
<td>589ns</td><td>368.717MB/s</td>
<td>830ns</td><td>261.847MB/s</td>
<td>272ns</td><td>799.403MB/s</td>
<td>536ns</td><td>425.437MB/s</td>
<td>535ns</td><td>425.937MB/s</td>
<td>908ns</td><td>251.018MB/s</td>
<td>2,753ns</td><td>82.8314MB/s</td>
<td>4,782ns</td><td>47.6763MB/s</td>
<td>60,590ns</td><td>3.76299MB/s</td>
</tr>
<tr>
<td>google_message1_proto3</td>
<td>738ns</td><td>294.517MB/s</td>
<td>947ns</td><td>229.116MB/s</td>
<td>463ns</td><td>469.982MB/s</td>
<td>638ns</td><td>357.597MB/s</td>
<td>602ns</td><td>378.568MB/s</td>
<td>1,029ns</td><td>221.676MB/s</td>
<td>2,779ns</td><td>82.0498MB/s</td>
<td>5,708ns</td><td>39.9467MB/s</td>
<td>60,357ns</td><td>3.77751MB/s</td>
</tr>
<tr>
<td>google_message2</td>
<td>291,595ns</td><td>277.242MB/s</td>
<td>232,824ns</td><td>347.611MB/s</td>
<td>101,818ns</td><td>793.67MB/s</td>
<td>202,914ns</td><td>416.778MB/s</td>
<td>201,577ns</td><td>419.543MB/s</td>
<td>230,345ns</td><td>367.145MB/s</td>
<td>350,244ns</td><td>241.46MB/s</td>
<td>1,181,602ns</td><td>71.5723MB/s</td>
<td>30,917,063ns</td><td>2.73538MB/s</td>
</tr>
<tr>
<td>google_message3_1</td>
<td>898,464,645ns</td><td>213.478MB/s</td>
<td>658,343,331ns</td><td>291.58MB/s</td>
<td>352,077,637ns</td><td>543.398MB/s</td>
<td>293,124,168ns</td><td>684.241MB/s</td>
<td>284,707,340ns</td><td>704.47MB/s</td>
<td>309,220,443ns</td><td>648.624MB/s</td>
<td>959,490,716ns</td><td>209.036MB/s</td>
<td>1,408,918,063ns</td><td>142.356MB/s</td>
<td>13,081,256,866ns</td><td>15.3324MB/s</td>
</tr>
<tr>
<td>google_message3_2</td>
<td>523,567ns</td><td>672.685MB/s</td>
<td>439,275ns</td><td>802.767MB/s</td>
<td>282,482ns</td><td>1.21505GB/s</td>
<td>231,784ns</td><td>1.54439GB/s</td>
<td>222,887ns</td><td>1.60603GB/s</td>
<td>268,253ns</td><td>1.33443GB/s</td>
<td>638,781ns</td><td>573.835MB/s</td>
<td>1,166,147ns</td><td>314.33MB/s</td>
<td>24,409,413ns</td><td>15.0169MB/s</td>
</tr>
<tr>
<td>google_message3_3</td>
<td>346ns</td><td>207.681MB/s</td>
<td>511ns</td><td>140.591MB/s</td>
<td>134ns</td><td>535.181MB/s</td>
<td>270ns</td><td>279.385MB/s</td>
<td>247ns</td><td>304.853MB/s</td>
<td>700ns</td><td>107.575MB/s</td>
<td>2,336ns</td><td>32.248MB/s</td>
<td>2,882ns</td><td>26.1431MB/s</td>
<td>28,590ns</td><td>2.63541MB/s</td>
</tr>
<tr>
<td>google_message3_4</td>
<td>1,883ns</td><td>7.96091GB/s</td>
<td>2,120ns</td><td>7.10024GB/s</td>
<td>1,611ns</td><td>9.3013GB/s</td>
<td>2,731ns</td><td>5.78006GB/s</td>
<td>2,697ns</td><td>5.85198GB/s</td>
<td>3,412ns</td><td>4.62609GB/s</td>
<td>6,322ns</td><td>2.49631GB/s</td>
<td>6,703ns</td><td>2.35442GB/s</td>
<td>20,150ns</td><td>802.061MB/s</td>
</tr>
<tr>
<td>google_message3_5</td>
<td>262ns</td><td>76.0072MB/s</td>
<td>386ns</td><td>51.6769MB/s</td>
<td>84ns</td><td>237.856MB/s</td>
<td>173ns</td><td>121.038MB/s</td>
<td>158ns</td><td>132.866MB/s</td>
<td>568ns</td><td>36.9197MB/s</td>
<td>2,018ns</td><td>10.3962MB/s</td>
<td>2,371ns</td><td>8.84659MB/s</td>
<td>16,757ns</td><td>1.25203MB/s</td>
</tr>
<tr>
<td>google_message4</td>
<td>52,237ns</td><td>331.46MB/s</td>
<td>43,929ns</td><td>404.862MB/s</td>
<td>41,750ns</td><td>427.99MB/s</td>
<td>30,825ns</td><td>606.228MB/s</td>
<td>31,720ns</td><td>589.13MB/s</td>
<td>35,213ns</td><td>530.692MB/s</td>
<td>61,161ns</td><td>305.543MB/s</td>
<td>106,886ns</td><td>174.834MB/s</td>
<td>2,376,051ns</td><td>7.86485MB/s</td>
</tr>
</table>

## <b>Serialization</b>
<table>
<tr>
<th rowspan="2"> </th>
<th colspan="2" rowspan="2">cpp</th>
<th colspan="6" rowspan="1">java</th>
<th colspan="6" rowspan="1">python</th>
</tr>
<tr>
<th colspan="2">serializeToByteArray</th>
<th colspan="2">serializeToByteString</th>
<th colspan="2">serializeToMemoryStream</th>
<th colspan="2">cpp-generated-code</th>
<th colspan="2">cpp-reflection</th>
<th colspan="2">pure-python</th>
</tr>
<tr>
<td>google_message1_proto2</td>
<td>152ns</td><td>1.39698GB/s</td>
<td>197ns</td><td>1.12915GB/s</td>
<td>196ns</td><td>1.13589GB/s</td>
<td>301ns</td><td>758.609MB/s</td>
<td>874ns</td><td>260.911MB/s</td>
<td>3,899ns</td><td>58.4815MB/s</td>
<td>39,458ns</td><td>5.77824MB/s</td>
</tr>
<tr>
<td>google_message1_proto3</td>
<td>219ns</td><td>959.305MB/s</td>
<td>193ns</td><td>1.15372GB/s</td>
<td>206ns</td><td>1.07824GB/s</td>
<td>284ns</td><td>802.337MB/s</td>
<td>952ns</td><td>239.4MB/s</td>
<td>6,779ns</td><td>33.6336MB/s</td>
<td>39,275ns</td><td>5.80524MB/s</td>
</tr>
<tr>
<td>google_message2</td>
<td>61,866ns</td><td>1.27429GB/s</td>
<td>81,739ns</td><td>1.01039GB/s</td>
<td>82,669ns</td><td>1022.99MB/s</td>
<td>105,880ns</td><td>798.736MB/s</td>
<td>84,845ns</td><td>996.755MB/s</td>
<td>1,459,107ns</td><td>57.9601MB/s</td>
<td>20,664,826ns</td><td>4.09246MB/s</td>
</tr>
<tr>
<td>google_message3_1</td>
<td>142,043,802ns</td><td>1.31916GB/s</td>
<td>202,287,972ns</td><td>991.496MB/s</td>
<td>233,128,141ns</td><td>860.332MB/s</td>
<td>302,570,212ns</td><td>662.88MB/s</td>
<td>131,785,828ns</td><td>1.48625GB/s</td>
<td>476,083,142ns</td><td>421.287MB/s</td>
<td>11,141,383,171ns</td><td>18.002MB/s</td>
</tr>
<tr>
<td>google_message3_2</td>
<td>157,404ns</td><td>2.15676GB/s</td>
<td>166,699ns</td><td>2.14736GB/s</td>
<td>171,985ns</td><td>2.08136GB/s</td>
<td>229,468ns</td><td>1.55997GB/s</td>
<td>149,402ns</td><td>2.39597GB/s</td>
<td>1,121,729ns</td><td>326.777MB/s</td>
<td>22,834,525ns</td><td>16.0527MB/s</td>
</tr>
<tr>
<td>google_message3_3</td>
<td>111ns</td><td>650.456MB/s</td>
<td>127ns</td><td>593.52MB/s</td>
<td>130ns</td><td>580.667MB/s</td>
<td>217ns</td><td>346.839MB/s</td>
<td>608ns</td><td>123.978MB/s</td>
<td>2,099ns</td><td>35.893MB/s</td>
<td>32,361ns</td><td>2.32834MB/s</td>
</tr>
<tr>
<td>google_message3_4</td>
<td>1,732ns</td><td>8.70154GB/s</td>
<td>2,681ns</td><td>5.88645GB/s</td>
<td>2,657ns</td><td>5.93946GB/s</td>
<td>6,458ns</td><td>2.44388GB/s</td>
<td>2,664ns</td><td>5.9241GB/s</td>
<td>3,889ns</td><td>4.05837GB/s</td>
<td>18,431ns</td><td>876.87MB/s</td>
</tr>
<tr>
<td>google_message3_5</td>
<td>81ns</td><td>246.33MB/s</td>
<td>74ns</td><td>283.278MB/s</td>
<td>81ns</td><td>259.167MB/s</td>
<td>102ns</td><td>206.37MB/s</td>
<td>567ns</td><td>37.0285MB/s</td>
<td>1,716ns</td><td>12.2228MB/s</td>
<td>17,514ns</td><td>1.1979MB/s</td>
</tr>
<tr>
<td>google_message4</td>
<td>11,130ns</td><td>1.56674GB/s</td>
<td>24,053ns</td><td>776.907MB/s</td>
<td>24,247ns</td><td>770.707MB/s</td>
<td>26,585ns</td><td>702.931MB/s</td>
<td>12,197ns</td><td>1.49623GB/s</td>
<td>91,106ns</td><td>205.116MB/s</td>
<td>2,091,638ns</td><td>8.93428MB/s</td>
</tr>
</table>

**Comment by BSBandme** [comment]
Any updates for this?

File: benchmarks/py_benchmark.py [hunk 2]
```diff
@@ -0,0 +1,124 @@
...
+               setup_method=None, full_iteration = 1):
+    self.full_iteration = full_iteration
+    self.module = module
+    self.test_method = test_method
+    self.setup_method = setup_method
+    
+  def set_test_method(self, test_method):
+    self.test_method = test_method
+  
+  def set_setup_method(self, setup_method):
```
Comment (anandolee) [inline benchmarks/py_benchmark.py]: set_setup_method and set_module is never used. Remove them maybe?

Comment (BSBandme) [reply benchmarks/py_benchmark.py]: Done

**Comment by anandolee** [comment]
LGTM

**Comment by BSBandme** [comment]
benchmarks/README.md  updated for python

File: benchmarks/README.md [hunk 3]
```diff
@@ -17,12 +17,23 @@ We are using [google/benchmark](https://github.com/google/benchmark) as the
...
 
-### JAVA
+### Java
 We're using maven to build the java benchmarks, which is the same as to build 
 the Java protobuf. There're no other tools need to install. We're using 
 [google/caliper](https://github.com/google/caliper) as benchmark tool, which 
 can be automaticly included by maven.
 
+### Python
+We're using python dev for testing the generated cpp proto version of python protobuf.
```
Comment (xfxyjwf) [inline benchmarks/README.md]: What is python dev?

File: m4/ax_python_devel.m4 [hunk 4]
```diff
@@ -0,0 +1,327 @@
+# ===========================================================================
```
Comment (xfxyjwf) [inline m4/ax_python_devel.m4]: Opensource protobuf python has been supporting python C++ build for a long time. Why do we suddenly need this now?

**Comment by BSBandme** [comment]
@xfxyjwf  any update?

File: benchmarks/README.md (hunk 3 above)
Comment (BSBandme) [reply benchmarks/README.md]: Fixed the readme description.

File: m4/ax_python_devel.m4 (hunk 4 above)
Comment (BSBandme) [reply m4/ax_python_devel.m4]: The python c++ build should be installed by python's module setuptools.setup. But for this benchmark, we need to use automake to build the python c++ extension by ourselves, so we need to configure the installed python header and lib directory before make.

**Comment by BSBandme** [comment]
Done changing to use pkg-config

File: configure.ac [hunk 5]
```diff
@@ -215,6 +215,7 @@ AX_CXX_COMPILE_STDCXX([11], [noext], [optional])
 #   too.
 export CFLAGS
 export CXXFLAGS
+
```
Comment (xfxyjwf) [inline configure.ac]: revert this file.

Comment (BSBandme) [reply configure.ac]: Done

**Pull request merged**